src/clj_pgp/message.clj

```diff
@@ -381,23 +406,74 @@
 
 ;; ## Reading PGP Messages
 
-(defn read-messages
-  "Reads message packets from an input source and returns a sequence of message
-  maps. Each message contains keys similar to the options used to build them,
-  describing the type of compression used, cipher encrypted with, etc. The
-  message content is stored in the `:data` entry.
+(defn- reduce-objects
+  "Reduces over the PGP objects the returns the resulting accumulator.
+  Verifys the integrity of each object and throws if its invalid."
+  [acc r-fn opts objects]
+  (reduce
+    (fn reduce-and-verify!
+      [acc message]
+      (reduce-message
+        message
+        acc
+        (fn [acc {:keys [object] :as message}]
+          ;; To be able to verify the integrity we must have consumed the stream itself.
+          ;; Make sure to call the reducing function and then verify the message.
+          (let [results (r-fn acc message)]
+            (when (and (isa? (type object) PGPEncryptedData)
```

Comment:
This should use `instance?`:
```suggestion
            (when (and (instance? PGPEncryptedData object)
```
 Unless you really want the consumer to be able to override your logic, use `class` instead of `type` anyway, since the latter is for _metadata_ typing:
```clojure
=> (type (vary-meta 'x assoc :type String))
#java/class java.lang.String
```